comptime.md

```diff
@@ -0,0 +1,125 @@
+# Compile-time execution
+
+# Summary 
+
+This RFC proposes to entirely change the current implementation of compile-time that will allow to run V language itself at comptime.
+
+# Motivation
+
+ The main purpose to do so is current limitation of compile-time evaluation:
+- No way to invoke arbitrary V functions
+- No way to choose function at compile time: 
+```v
+$if flag_a {
+    fn foo() {
+        println('bar')
+    }
+} $else {
+    fn foo() {
+        println('foo')
+    }
+}
+```
+^ Compiler cannot handle this code 
+
+- It limits our generics and compile-time reflection. With proper comptime it would be possible to generate complex serialization/deserialization functions without compiler support. 
+- Implementing proper comptime in fact would simplify our compiler pipelines, at the moment there is a lot of edge-cases for handling current limited comptime. 
+- Proper metaprogramming is must have in every modern language 
+
+
+# Guide-level explanation 
+
+With this RFC the main change is that all expressions and statements prefixed with `$` will be treated as compile-time and all of them can be executed *before* codegen. Current compile-time is quite limited and allows only a few statements (`$if`,`$for`) and only specific calls.
+
+You can compare new way of handling comtime to Zig comptime functionality. It would allow us not only to perform some computations at compile-time but also generate new structures and functions at compile-time. All generic types would not be just some compiler magic but `v.ast.Type` and you should be able to invoke all of its methods at compile-time, get type symbol from it and even create new `v.ast.Type`s at compile-time.  
+
+First where comptime can be used is automatic generation of serialization/deserialization functions. Here's small example of how we can implement automatic `str()` for types: 
+```v
+fn to_str<T>(val T) string {
+    // FIXME: Decide if generic type is `v.ast.Type` or `v.ast.TypeSymbol`
```

Comment:
@danieldaeschle I think setting generic comptime variable types to `v.ast.TypeSymbol` would be better and would make usage simpler, otherwise we have to provide some builtin functions or special syntax to access TypeSymbol information and I think it is not the best option